x-pack/plugins/painless_lab/public/application/context.tsx

```diff
@@ -8,50 +8,89 @@ import React, { createContext, ReactNode, useState, useContext } from 'react';
 import { HttpSetup } from 'src/core/public';
 
 import { Links } from '../links';
+import { exampleScript, painlessContextOptions } from './constants';
+import { Store, Payload, Validation } from './types';
 
-import { initialState, Store } from './store';
-import { PAINLESS_LAB_KEY } from './constants';
+interface AppContextProviderArgs {
+  children: ReactNode;
+  value: {
+    http: HttpSetup;
+    links: Links;
+  };
+}
 
 interface ContextValue {
-  state: Store;
-  updateState: (nextState: (s: Store) => Partial<Store>) => void;
+  store: Store;
+  updatePayload: (changes: Partial<Payload>) => void;
   services: {
     http: HttpSetup;
   };
   links: Links;
 }
 
+const initialPayload = {
+  context: painlessContextOptions[0].value,
+  code: exampleScript,
+  parameters: `{
+  "string-parameter": "yay",
+  "number-parameter": 1.5,
+  "boolean-parameter": true
+}`,
+  index: 'default-index',
+  document: `{
+  "my-field": "field-value"
+}`,
+  query: '',
+};
+
 const AppContext = createContext<ContextValue>(undefined as any);
 
-interface AppContextProviderArgs {
-  children: ReactNode;
-  value: {
-    http: HttpSetup;
-    links: Links;
+const validatePayload = (payload: Payload): Validation => {
+  const { index } = payload;
+
+  // For now just validate that the user has entered an index.
+  const indexExists = Boolean(index || index.trim());
+
+  return {
+    isValid: indexExists,
+    fields: {
+      index: indexExists,
+    },
   };
-}
+};
 
 export const AppContextProvider = ({
   children,
   value: { http, links },
 }: AppContextProviderArgs) => {
-  const [state, setState] = useState<Store>(() => ({
-    ...initialState,
+  const PAINLESS_LAB_KEY = 'painlessLabState';
+
+  const defaultPayload = {
+    ...initialPayload,
     ...JSON.parse(localStorage.getItem(PAINLESS_LAB_KEY) || '{}'),
-  }));
+  };
 
-  const updateState = (getNextState: (s: Store) => Partial<Store>): void => {
```

Comment:
@jloleysens Was there a particular reason to accept a callback here instead of the state change?

Reply:
The primary reason was to provide callers of `updateState` with a reference to the previous state so they can make decisions about it if needed, e.g., `updateState(s => s.something ? ...  : ...)` and so that it is a bit more convenient to work with as the store grows in complexity (otherwise we have to fetch the store from `const { store } = useAppContext()` and pass use it before we update something. So it is primarily a mechanism of optimisation.

This followed the `useState` pattern, but I am happy to have it accept both a callback and an object :+1:

Reply:
OK, thanks for explaining! I follow your reasoning until this part:

> otherwise we have to fetch the store from const { store } = useAppContext() and use it when we update something and then update it again in the future if the shape changes.

This is where I get lost. Is there an example of this in the codebase? Philosophically I would suggest waiting until this becomes a problem before introducing a solution for it.